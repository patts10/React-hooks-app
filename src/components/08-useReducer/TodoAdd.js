import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

  const [ { description }, handleInputChange, reset ] = useForm({
    description: ''
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };


    handleAddTodo( newTodo );

    reset();
  }

  return (
    <>
      <h4>Add TODO</h4>
          <hr />

          <form onSubmit={ handleSubmit }>
            <input 
              type="text"
              name="description"
              placeholder="Learn..."
              autoComplete="off"
              value={ description }
              onChange={ handleInputChange }
              className="form-control block w-full px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
            />

            <button
              className="w-full mt-1 px-6 py-2.5 bg-white text-blue-600 hover:text-white font-medium text-xs border border-blue-600 leading-tight rounded shadow-md hover:bg-blue-700 hover:shadow-lg focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
              type="submit"
            >
              Add
            </button>
          </form>
    </>
  )
}
