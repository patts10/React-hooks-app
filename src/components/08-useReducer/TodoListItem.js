import React from "react";

export const TodoListItem = ({ todo, index, handleDelete, handleToggle }) => {
  return (
    <li className="flex items-center justify-between px-6 py-2 w-full">
      <p
        className={ todo.done ? "line-through cursor-pointer mx-0" : "cursor-pointer mx-0" }
        onClick={() => handleToggle(todo.id)}
      >
        {index + 1}. {todo.desc}
      </p>
      <button
        className="bg-red-500 shadow-lg shadow-red-500/50 rounded py-1 px-2 mt-3 text-white"
        onClick={() => handleDelete(todo.id)}
      >
        Delete
      </button>
    </li>
  );
};
