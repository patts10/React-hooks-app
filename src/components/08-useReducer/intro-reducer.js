
const initialState = [{
  id:1,
  todo: 'Buy bread',
  done: false
}];

const todoReducer = ( state = initialState, action ) => {

  if (action?.type === 'add') {
    return [...state, action.payload]
  }

  
  return state;
}

let todos = todoReducer();

const newTodo = {
  id:2,
  todo: 'Buy milk',
  done: false
};


const agregarTodoAction = {
  type: 'add',
  payload: newTodo
}

todos = todoReducer( todos, agregarTodoAction );






console.log(todos);