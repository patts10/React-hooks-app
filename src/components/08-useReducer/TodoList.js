import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
  return (
    <ul className="bg-white rounded-lg w-full text-gray-900">
            {todos.map((todo, i) => (
              
              <TodoListItem 
                key={ todo.id } 
                todo={ todo } 
                index={ i } 
                handleDelete={ handleDelete } 
                handleToggle={ handleToggle } 
              />
            ))}
          </ul>
  )
}
