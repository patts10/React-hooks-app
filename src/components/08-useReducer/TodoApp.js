import React, { useEffect, useReducer } from "react";
import "./styles.css";
import { TodoAdd } from "./TodoAdd";
import { TodoList } from "./TodoList";
import { todoReducer } from "./todoReducer";

const init = () => {

  return JSON.parse(localStorage.getItem('todo')) || [];
};

export const TodoApp = () => {

  const [ todos, dispatch ] = useReducer(todoReducer, [], init);

  useEffect( () => {

    localStorage.setItem('todo', JSON.stringify(todos));

  }, [ todos ]);

  const handleDelete = ( id ) => {
    const action = {
      type: 'delete',
      payload: id
    }

    dispatch(action);
  };

  const handleToggle = ( todoId ) => {
    dispatch({
      type: 'toggle',
      payload: todoId
    });
  };

  const handleAddTodo = ( newTodo ) => {

    dispatch( {
      type: 'add',
      payload: newTodo
    } );
  }


  return (
    <div>
      <h1>TodoApp ( {todos.length} )</h1>
      <hr />

      <div className="flex justify-center">
        <div className="w-7/12">
          
          <TodoList 
            todos={ todos } 
            handleDelete={ handleDelete } 
            handleToggle={ handleToggle }
          />
        </div>

        <div className="w-5/12">
          <TodoAdd handleAddTodo={ handleAddTodo } />
        </div>
      </div>
    </div>
  );
};
