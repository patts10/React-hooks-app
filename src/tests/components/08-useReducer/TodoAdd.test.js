import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";
import { shallow } from "enzyme";

describe("Tests on <TodoAdd />", () => {
  const handleAddTodo = jest.fn();

  const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

  test("should show correctly", () => {

    expect(wrapper).toMatchSnapshot();
    
  });

  test('should not call handleAddTodo function', () => { 
    
    const formSubmit = wrapper.find('form').prop('onSubmit');

    formSubmit({ preventDefault(){} });

    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test('should call handleAddTodo function', () => { 
    
    const value = 'Learn Vuejs';
    
    wrapper.find('input').simulate('change', {
      target: {
        value,
        name: 'description'
      }
    })
    
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault(){} })

    expect(handleAddTodo).toHaveBeenCalled();
    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    expect(handleAddTodo).toHaveBeenCalledWith( expect.any(Object) );
    expect( handleAddTodo ).toHaveBeenCalledWith({
      id: expect.any( Number ),
      desc: value,
      done: false
    });
    expect(wrapper.find('input').prop('value')).toBe('');

   });

});
