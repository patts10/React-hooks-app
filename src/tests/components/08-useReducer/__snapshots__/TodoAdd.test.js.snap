// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <TodoAdd /> should show correctly 1`] = `
<Fragment>
  <h4>
    Add TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control block w-full px-3 py-1.5 text-base font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
      name="description"
      onChange={[Function]}
      placeholder="Learn..."
      type="text"
      value=""
    />
    <button
      className="w-full mt-1 px-6 py-2.5 bg-white text-blue-600 hover:text-white font-medium text-xs border border-blue-600 leading-tight rounded shadow-md hover:bg-blue-700 hover:shadow-lg focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
